### ISSUE TITLE
Possible split-brain of Akka Cluster using K8s Deployment

### ISSUE THREAD

**Issue by notxcain** [akka/akka-management#156]
I want to describe a possible split-brain scenario (even with SBR enabled) when using K8s Deployment for Akka Cluster.

## How `Deployment` works
1. `Deployment` assumes it's replicas are stateless and that it's safe to have more of them than configured by `replicas` parameter
2. Kubernetes controller manager will try to check `nodeStatusUpdateRetry` times every `--node-monitor-period` of time. After `--node-monitor-grace-period` it will consider node unhealthy. It will remove its pods after `--pod-eviction-timeout`

**Comment by notxcain** [comment]
Ooops, accidentally hit `Ctrl+Enter` - WIP

**Comment by ktoso** [comment]
Hi @notxcain, your writeup is a bit weird... specifically:

> Eventually K8s starts 3 containers on reachable nodes, they form cluster. Now we have 2 clusters.

Why 3 new ones? 
Why would they form a cluster?

>What about StatefulSet?

Why not stateful set is documented here: https://github.com/akka/akka-management/issues/134
Please give it a look and correct us if you think that analysis is incorrect.

**Comment by notxcain** [comment]
> Why 3 new ones?
One member left on reachable worker will kill itself. K8s will restart it. Also k8s will start evicted pods on reachable nodes. At this moment there will be 5 running members.

> Why would they form a cluster?
They would form a cluster because of how cluster-bootstrap works. Members left on unreachable nodes won't be listed in headless service DNS records, so there gonna be no evidence of their existence.

> Please give it a look and correct us if you think that analysis is incorrect.
I'll take a closer look later.

**Comment by ktoso** [comment]
> One member left on reachable worker will kill itself. K8s will restart it. Also k8s will start evicted pods on reachable nodes. At this moment there will be 5 running members.

I have a feeling wording is not quite precise here... Could you please explain the exact steps we are worried about and how it relates to "akka nodes" "pods" "physical nodes" etc?

>      They would form a cluster because of how cluster-bootstrap works. 

Again I feel you're making a huge mental leap over all the details between. Which nodes and for what time are visible in discovery via k8s APIs etc plays a huge role in these things. I don't see these taken into consideration in this ticket so far?

**Comment by notxcain** [comment]
> I have a feeling wording is not quite precise here... Could you please explain the exact steps we are worried about and how it relates to "akka nodes" "pods" "physical nodes" etc?

akka member == pod
worker == physical node

1. Workers running 2 of 3 members become unreachable from the point of view of K8s and the 3rd member.
2. K8s evicts pods from unreachable workers and schedule them on reachable, also removes unreachable pods from headless service DNS record
3. Two evicted members are started, so there are 3 member on reachable workers and 2 still running on unreachable worker.

**Comment by hseeberger** [comment]
I totally agree that this could happen: SBR and k8s don't have the same understanding about which nodes (they even don't have the same understanding what a node is, but in the case of a network partition that doesn't matter) are available or not.

In a three node (physical) cluster k8s might consider nodes A and B unreachable. Assuming member nodes (ActorSystem) a, b and c are running on (within a pod) on each of the nodes A, B and C, SBR on c detects a and b unreachable whereas SBR on a and b detects c as unreachable. 

Hence only k8s and SBR on c are in sync. But with quorum or majority, SBR will down c and hence leave a and b up which unfortunately run on A and B which are considered unreachable by k8s.

In the words of Tyrion Lannister: "we're fucked"

**Comment by longshorej** [comment]
I'd like to draw attention to the "join existing" flag (see reference.conf) which can help here. The initial deployment of your Akka Cluster app needs to have this flag disabled, but then any future rolling updates can enable it. These are the majority of cases of course, so in practice it shouldn't introduce much operator overhead.

**Comment by hseeberger** [comment]
@longshorej are you referring to `akka.management.cluster.bootstrap.form-new-cluster`?

**Comment by longshorej** [comment]
Yeah -- oops, I forgot it was named that. So, you'd need that set to true for the initial deployment. Subsequent deployments can disable it and they'll then only join an existing cluster. You'd need appropriate readiness checks defined and https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#max-unavailable set to 10 or so, depending upon cluster size.

**Comment by thomschke** [comment]
Try to poll the current replicas (`.status.readyReplicas`) from api-server and compute the matching `quorum-size` incl. recommended `stable-after` [(see table)](https://developer.lightbend.com/docs/akka-commercial-addons/current/split-brain-resolver.html#strategies).

AND 

Stop your SBR as long as if:
- the api-server is reachable but there is running an deployment-update (`.status.replicas != .status.readyReplicas`) or
- the api-server is temporary unreachable (caused by k8s cluster update or crash)

Sure, there is not a “one size fits all” solution to this problem. But I think this strategy fits the characteristics of a k8s cluster.

**Comment by thomschke** [comment]
Currently all members are used during bootstrapping -> #254 
That's why new cluster nodes tries to join unreachable nodes or nodes in status "weaklyUp" ...
